src/framework/jobState/LocalStepJobState/LocalStepJobState.ts

```diff
@@ -0,0 +1,97 @@
+import { Promisable } from 'type-fest';
+
+import { Entity, Relationship } from '../../types';
+import { JobState, GraphObjectFilter, GraphObjectIteratee } from '../types';
+
+import { flushDataToDisk } from './flushDataToDisk';
+
+export const GRAPH_OBJECT_BUFFER_THRESHOLD = 500; // arbitrarily selected, subject to tuning
+
+interface LocalStepJobStateInput {
+  step: string;
+  cacheDirectory?: string;
+}
+
+export class LocalStepJobState implements JobState {
+  readonly cacheDirectory?: string;
+  readonly step: string;
+
+  entities: Entity[];
+  relationships: Relationship[];
+
+  constructor({ step, cacheDirectory }: LocalStepJobStateInput) {
+    this.step = step;
+    this.cacheDirectory = cacheDirectory;
+
+    this.entities = [];
+    this.relationships = [];
+  }
+
+  async addEntities(newEntities: Entity[]) {
+    this.entities = this.entities.concat(newEntities);
+
+    if (this.entities.length >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
+      await this.flushEntitiesToDisk();
+    }
+  }
+
+  async addRelationships(newRelationships: Relationship[]) {
+    this.relationships = this.relationships.concat(newRelationships);
+
+    if (this.relationships.length >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
+      await this.flushRelationshipsToDisk();
+    }
+  }
+
+  async iterateEntities(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<Entity>,
+  ) {
+    /* stub for now */
+    return;
+  }
+
+  async iterateRelationships(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<Relationship>,
+  ) {
+    /* stub for now */
+    return;
+  }
+
+  async flush() {
+    const work: Promisable<void>[] = [];
+
+    if (this.entities.length) {
+      work.push(this.flushEntitiesToDisk());
+    }
+
+    if (this.relationships.length) {
+      work.push(this.flushRelationshipsToDisk());
+    }
+
+    await Promise.all(work);
```

Comment:
I read the docs at https://github.com/sindresorhus/type-fest/blob/HEAD/source/promisable.d.ts, but I don't think I fully understand the value in this scenario. Would you mind helping me? 😄

I think it would be good to have `flushEntitiesToDisk()` worry about whether there is anything to flush or not, or else that function cannot be called without the caller verifying that it can be called.

```suggestion
    await Promise.all([
      this.flushEntitiesToDisk(),
      this.flushRelationshipsToDisk(),
    ]);
```

Reply:
tbh, idk why I didn't just use `Promise` instead of `Promisable`. Removing.

Reply:
And I'm cool with moving the check into `flushEntitiesToDisk`